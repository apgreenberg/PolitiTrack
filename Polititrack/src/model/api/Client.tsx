import { ApolloClient, InMemoryCache } from '@apollo/client';

import OPENSTATES_OAUTH_TOKEN from '../../types/env';

const client = new ApolloClient({
  uri: 'https://openstates.org/graphql',
  headers: {
    'X-API-KEY': `${OPENSTATES_OAUTH_TOKEN}`,
  },
  cache: new InMemoryCache(),
});

export default client;
