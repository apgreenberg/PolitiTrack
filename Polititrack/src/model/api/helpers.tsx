export enum Status {
  Loading,
  Error,
  Success,
}
