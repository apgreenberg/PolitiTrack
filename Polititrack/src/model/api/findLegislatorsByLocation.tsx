import { gql, useQuery } from '@apollo/client';
import { Status } from './helpers';

export interface LegislatorSearchResults {
  status: Status;
  errorMessage: string;
  data: LegislatorSearchNode[];
}

export interface LegislatorSearchNode {
  id: string;
  name: string;
}

export const FIND_LEGISLATORS_BY_LOCATION_QUERY = gql`
  query findLegislatorByLocation($latitude: Float!, $longitude: Float!) {
    people(latitude: $latitude, longitude: $longitude, first: 10) {
      edges {
        node {
          id
          name
        }
      }
    }
  }
`;

/**
 * Gets the user's legislators based on the location of their device
 * @param requestParams - the request parameters (latitude, longitude)
 * @returns - the status of the request (loading, error or success) and the user's legislators on success
 */
export function findLegislatorsByLocation(
  requestParams: string[]
): LegislatorSearchResults {
  const legislatorSearchResults: LegislatorSearchResults = {
    status: Status.Loading,
    errorMessage: '',
    data: [],
  };
  const { loading, error, data } = useQuery(
    FIND_LEGISLATORS_BY_LOCATION_QUERY,
    {
      variables: {
        latitude: parseFloat(requestParams[0]),
        longitude: parseFloat(requestParams[1]),
      },
    }
  );

  if (loading) {
    return legislatorSearchResults;
  }
  if (error) {
    legislatorSearchResults.status = Status.Error;
    legislatorSearchResults.errorMessage = error.message;
    return legislatorSearchResults;
  }
  legislatorSearchResults.status = Status.Success;
  let wrapper: {
    node: {
      id: '';
      name: '';
    };
  };
  for (wrapper of data.people.edges) {
    legislatorSearchResults.data.push({
      id: wrapper.node.id,
      name: wrapper.node.name,
    });
  }
  return legislatorSearchResults;
}
