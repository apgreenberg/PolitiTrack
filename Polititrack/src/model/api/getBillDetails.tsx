import { gql, useQuery } from '@apollo/client';
import { Status } from './helpers';
import axios, { AxiosResponse } from 'axios';
import OPENSTATES_OAUTH_TOKEN from '../../types/env';

export interface BillDetails {
  status: Status;
  errorMessage: string;
  data: BillDetailNode;
}

export interface BillDetailNode {
  title: string;
  from: string;
  openstatesUrl: string;
  updatedAt: string;
}

export interface GetBillDetailsResponse {
  data: {
    bill: {
      fromOrganization: {
        name: string;
      };
      title: string;
      openstatesUrl: string;
      updatedAt: string;
    };
  };
}

export const BILL_DETAILS_QUERY = `
  query getBillDetails($id: String!) {
    bill(id: $id) {
      title
      openstatesUrl
      updatedAt
      fromOrganization {
        name
      }
    }
  }
`;

/**
 * Gets the bill details
 * @param requestParams - the request parameters (bill id)
 * @returns - the status of the request (loading, error or success) and the bill details on success
 */
export function getBillDetails(requestParams: string[]): BillDetails {
  const billDetails: BillDetails = {
    status: Status.Loading,
    errorMessage: '',
    data: { title: '', from: '', openstatesUrl: '', updatedAt: '' },
  };
  const { loading, error, data } = useQuery(
    gql`
      ${BILL_DETAILS_QUERY}
    `,
    {
      variables: {
        id: requestParams[0],
      },
    }
  );

  if (loading) {
    return billDetails;
  }
  if (error) {
    billDetails.status = Status.Error;
    billDetails.errorMessage = error.message;
    return billDetails;
  }
  billDetails.status = Status.Success;
  billDetails.data.title = data.bill.title;
  billDetails.data.openstatesUrl = data.bill.openstatesUrl;
  billDetails.data.updatedAt = data.bill.updatedAt;
  billDetails.data.from = data.bill.fromOrganization.name;
  return billDetails;
}

/**
 * Gets the bill details but using Axios instead of Apollo hooks
 * Necessary for background updates
 * @param requestParams - the request parameters (bill id)
 * @returns - the status of the request (loading, error or success) and the bill details on success
 */
export async function getBillDetailsAxios(
  requestParams: string[]
): Promise<BillDetails> {
  const billDetails: BillDetails = {
    status: Status.Loading,
    errorMessage: '',
    data: { title: '', from: '', openstatesUrl: '', updatedAt: '' },
  };
  try {
    const response: AxiosResponse<GetBillDetailsResponse> = await axios({
      url: 'https://openstates.org/graphql',
      method: 'post',
      headers: {
        'X-API-KEY': `${OPENSTATES_OAUTH_TOKEN}`,
      },
      data: {
        query: BILL_DETAILS_QUERY,
        variables: {
          id: requestParams[0],
        },
      },
    });
    billDetails.status = Status.Success;
    billDetails.data.title = response.data.data.bill.title;
    billDetails.data.openstatesUrl = response.data.data.bill.openstatesUrl;
    billDetails.data.updatedAt = response.data.data.bill.updatedAt;
    billDetails.data.from = response.data.data.bill.fromOrganization.name;
    return billDetails;
  } catch {
    (error: Error) => {
      billDetails.status = Status.Error;
      billDetails.errorMessage = error.message;
      return billDetails;
    };
  }
  return billDetails;
}
