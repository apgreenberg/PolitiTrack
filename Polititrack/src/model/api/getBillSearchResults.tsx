import { gql, useQuery } from '@apollo/client';
import { Status } from './helpers';

export interface BillSearchResults {
  status: Status;
  errorMessage: string;
  jurisdiction: string;
  billType: string;
  data: BillSearchNode[];
}

export interface BillSearchNode {
  id: string;
  name: string;
}

export const BILL_SEARCH_QUERY = gql`
  query getBillSearch(
    $jurisdiction: String!
    $session: String!
    $searchQuery: String!
  ) {
    bills(
      jurisdiction: $jurisdiction
      session: $session
      searchQuery: $searchQuery
      first: 30
    ) {
      edges {
        node {
          title
          id
        }
      }
    }
  }
`;

/**
 * Gets the search result bills
 * @param requestParams - the request
 *        legislature - the legislature/jurisdiction to search
 *        session - the legislative session to search
 *        billType - the bill subject that the user wants to lookup
 * @returns - a list of bills meeting the search criteria
 */
export function getBillSearchResults(
  requestParams: string[]
): BillSearchResults {
  const billSearchResults: BillSearchResults = {
    status: Status.Loading,
    errorMessage: '',
    jurisdiction: requestParams[0],
    billType: requestParams[2],
    data: [],
  };
  const { loading, error, data } = useQuery(BILL_SEARCH_QUERY, {
    variables: {
      jurisdiction: requestParams[0],
      session: requestParams[1],
      searchQuery: requestParams[2],
    },
  });

  if (loading) {
    return billSearchResults;
  }
  if (error) {
    billSearchResults.status = Status.Error;
    billSearchResults.errorMessage = error.message;
    return billSearchResults;
  }
  billSearchResults.status = Status.Success;
  let wrapper: {
    node: {
      id: '';
      title: '';
    };
  };
  for (wrapper of data.bills.edges) {
    billSearchResults.data.push({
      id: wrapper.node.id,
      name: wrapper.node.title,
    });
  }
  return billSearchResults;
}
