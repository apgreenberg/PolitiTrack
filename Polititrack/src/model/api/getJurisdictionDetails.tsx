import { gql, useQuery } from '@apollo/client';
import { Status } from './helpers';
import axios, { AxiosResponse } from 'axios';
import OPENSTATES_OAUTH_TOKEN from '../../types/env';

export interface JurisdictionDetails {
  status: Status;
  errorMessage: string;
  data: JurisdictionDetailNode;
}

export interface JurisdictionDetailNode {
  identifier: string;
  name: string;
  inSession: boolean;
  startDate: string;
  endDate: string;
}

export interface GetJurisdictionDetailsResponse {
  data: {
    jurisdiction: {
      legislativeSessions: {
        edges: GetJurisdictionDetailsResponseNode[];
      };
    };
  };
}

export interface GetJurisdictionDetailsResponseNode {
  node: {
    identifier: string;
    name: string;
    startDate: string;
    endDate: string;
  };
}

export const JURISDICTION_DETAILS_QUERY = `
  query getJurisdictionDetails($id: String!) {
    jurisdiction(id: $id) {
      legislativeSessions(first: 30) {
        edges {
          node {
            identifier
            name
            startDate
            endDate
          }
        }
      }
    }
  }
`;

/**
 * Gets the jurisdiction details
 * @param requestParams - the request parameters (jurisdiction id)
 * @returns - the status of the request (loading, error or success) and the jurisdiction details on success
 */
export function getJurisdictionDetails(
  requestParams: string[]
): JurisdictionDetails {
  const { loading, error, data } = useQuery(
    gql`
      ${JURISDICTION_DETAILS_QUERY}
    `,
    {
      variables: { id: requestParams[0] },
    }
  );

  const jurisdictionDetails: JurisdictionDetails = {
    status: Status.Loading,
    errorMessage: '',
    data: {
      identifier: '',
      name: 'Currently not in session',
      inSession: false,
      startDate: 'NA',
      endDate: 'NA',
    },
  };

  if (loading) {
    return jurisdictionDetails;
  }
  if (error) {
    jurisdictionDetails.status = Status.Error;
    jurisdictionDetails.errorMessage = error.message;
    return jurisdictionDetails;
  }
  jurisdictionDetails.status = Status.Success;
  findCurrentSession(
    jurisdictionDetails,
    data.jurisdiction.legislativeSessions.edges
  );
  return jurisdictionDetails;
}

/**
 * Same as getJurisdictionDetails, but uses Axios instead of Apollo hooks
 * Necessary for background updates
 * @param requestParams - the request parameters (jurisdiction id)
 * @returns - the status of the request (error or success) and the jurisdiction details on success
 */
export async function getJurisdictionDetailsAxios(
  requestParams: string[]
): Promise<JurisdictionDetails> {
  const jurisdictionDetails: JurisdictionDetails = {
    status: Status.Loading,
    errorMessage: '',
    data: {
      identifier: '',
      name: 'Currently not in session',
      inSession: false,
      startDate: 'NA',
      endDate: 'NA',
    },
  };
  try {
    const response: AxiosResponse<GetJurisdictionDetailsResponse> = await axios(
      {
        url: 'https://openstates.org/graphql',
        method: 'post',
        headers: {
          'X-API-KEY': `${OPENSTATES_OAUTH_TOKEN}`,
        },
        data: {
          query: JURISDICTION_DETAILS_QUERY,
          variables: {
            id: requestParams[0],
          },
        },
      }
    );
    jurisdictionDetails.status = Status.Success;
    findCurrentSession(
      jurisdictionDetails,
      response.data.data.jurisdiction.legislativeSessions.edges
    );
    return jurisdictionDetails;
  } catch {
    (error: Error) => {
      jurisdictionDetails.status = Status.Error;
      jurisdictionDetails.errorMessage = error.message;
      return jurisdictionDetails;
    };
  }
  return jurisdictionDetails;
}

/**
 * Given an array of legislative sessions, finds the current one if possible
 * @param jurisdictionDetails - the current jurisidiction details
 * @param sessions - the possible legislative sessions
 */
function findCurrentSession(
  jurisdictionDetails: JurisdictionDetails,
  sessions: GetJurisdictionDetailsResponseNode[]
): void {
  let wrapper = {
    node: {
      identifier: '',
      name: '',
      startDate: '',
      endDate: '',
    },
  };
  for (wrapper of sessions) {
    if (Date.parse(wrapper.node.endDate) > Date.parse(Date())) {
      jurisdictionDetails.data.identifier = wrapper.node.identifier;
      jurisdictionDetails.data.name = wrapper.node.name;
      jurisdictionDetails.data.inSession = true;
      jurisdictionDetails.data.startDate = wrapper.node.startDate;
      jurisdictionDetails.data.endDate = wrapper.node.endDate;
      break;
    }
  }
}
