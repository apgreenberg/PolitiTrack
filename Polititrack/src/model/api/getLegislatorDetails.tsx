import { gql, useQuery } from '@apollo/client';
import { Status } from './helpers';

export interface LegislatorDetails {
  status: Status;
  errorMessage: string;
  data: LegislatorDetailNode;
}

export interface LegislatorDetailNode {
  email: string;
  image: string;
  party: string;
}

export const LEGISLATOR_DETAILS_QUERY = gql`
  query getLegislatorDetails($id: ID!) {
    person(id: $id) {
      email
      image
      primaryParty
    }
  }
`;

/**
 * Gets the legislator details
 * @param requestParams - the request parameters (legislator id)
 * @returns - the status of the request (loading, error or success) and the legislator details on success
 */
export function getLegislatorDetails(
  requestParams: string[]
): LegislatorDetails {
  const legislatorDetails: LegislatorDetails = {
    status: Status.Loading,
    errorMessage: '',
    data: { email: '', image: '', party: '' },
  };
  const { loading, error, data } = useQuery(LEGISLATOR_DETAILS_QUERY, {
    variables: {
      id: requestParams[0],
    },
  });

  if (loading) {
    return legislatorDetails;
  }
  if (error) {
    legislatorDetails.status = Status.Error;
    legislatorDetails.errorMessage = error.message;
    return legislatorDetails;
  }
  legislatorDetails.status = Status.Success;
  legislatorDetails.data.email = data.person.email;
  legislatorDetails.data.image = data.person.image;
  legislatorDetails.data.party = data.person.primaryParty;
  return legislatorDetails;
}
