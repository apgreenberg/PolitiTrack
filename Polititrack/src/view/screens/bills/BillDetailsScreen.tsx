import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  Share,
  Button,
} from 'react-native';
import { ScrollView } from 'react-native-gesture-handler';
import * as WebBrowser from 'expo-web-browser';
import { StackScreenProps } from '@react-navigation/stack';
import { StackParams } from '../../../navigation/StackParams';

import { BillDetails, getBillDetails } from '../../../model/api/getBillDetails';
import { updateLastUpdated } from '../../../model/database/handleContent';
import { Status } from '../../../model/api/helpers';

import CustomButton from '../../components/CustomButton';
import LoadingScreen from '../LoadingScreen';
import ErrorScreen from '../ErrorScreen';

import text from '../../styles/text';

type Props = StackScreenProps<StackParams, 'Bill Details'>;

/**
 * Displays the bill details for a given bill id
 * @param - the state to request the bill details for
 * @returns - the functional components representing the screen
 */
const BillDetailsScreen: React.FC<Props> = ({ route, navigation }) => {
  React.useEffect(() => {
    (async () => updateLastUpdated(route.params.content))();
  }, []);

  const onShare = async (message: string, url: string): Promise<void> => {
    try {
      await Share.share({
        message: message + ' ' + url,
      });
    } catch (error) {
      alert(error.message);
    }
  };
  const billDetails: BillDetails = getBillDetails([route.params.content.id]);

  if (billDetails.status === Status.Loading) {
    return <LoadingScreen />;
  }
  if (billDetails.status === Status.Error) {
    return <ErrorScreen message={billDetails.errorMessage} />;
  }
  return (
    <SafeAreaView style={{ flex: 1 }}>
      <ScrollView style={{ flex: 1, paddingBottom: 100 }}>
        <Text style={text.largeText}>{billDetails.data.title}</Text>
        <View style={{ borderBottomWidth: 2, padding: 5 }}>
          <Text style={text.mediumText}>
            Introduced in: {billDetails.data.from}
          </Text>
        </View>
        <View style={styles.buttonPanelContainer}>
          <View style={styles.buttonContainer}>
            <CustomButton
              type="Abstract"
              destination="Bill Abstract"
              params={route.params.content}
            ></CustomButton>
          </View>
          <View style={styles.buttonContainer}>
            <CustomButton
              type="Sponsors"
              destination="Bill Sponsors"
              params={route.params.content}
            ></CustomButton>
          </View>
          <View style={styles.buttonContainer}>
            <CustomButton
              type="Vote History"
              destination="Bill Vote History"
              params={route.params.content}
            ></CustomButton>
          </View>
          <View style={styles.buttonContainer}>
            <CustomButton
              type="Versions"
              destination="Bill Versions"
              params={route.params.content}
            ></CustomButton>
          </View>
          <View style={{ marginTop: 20 }}>
            <Button
              onPress={() =>
                onShare(billDetails.data.title, billDetails.data.openstatesUrl)
              }
              title="Share"
            />
          </View>
        </View>
        <View style={{ borderBottomWidth: 2, padding: 10 }}>
          <Text
            style={[text.mediumText, { color: 'blue' }]}
            onPress={() =>
              WebBrowser.openBrowserAsync(billDetails.data.openstatesUrl)
            }
          >
            {billDetails.data.openstatesUrl}
          </Text>
          <Text style={text.mediumText}>
            Updated at {billDetails.data.updatedAt}
          </Text>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  buttonContainer: {
    alignItems: 'center',
    height: '25%',
    width: '100%',
  },
  buttonPanelContainer: {
    padding: 20,
    justifyContent: 'space-around',
    height: 350,
  },
});

export default BillDetailsScreen;
