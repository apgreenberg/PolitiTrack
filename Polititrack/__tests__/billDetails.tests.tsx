import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { GraphQLError } from 'graphql';

import { wait, billProps } from './common';
import BillDetailsScreen from '../src/view/screens/bills/BillDetailsScreen';
import { BILL_DETAILS_QUERY } from '../src/model/api/getBillDetails';
import gql from 'graphql-tag';

/**
 * Tests the Bill Details Screen
 */
describe('Bill Detail tests', () => {
  test('bill loading and success screen', async () => {
    const billDetailsMockSuccess = {
      request: {
        query: gql`
          ${BILL_DETAILS_QUERY}
        `,
        variables: {
          id: 'ocd-bill/2bd26615-0165-4f07-a53e-b9739f53dae6',
        },
      },
      result: {
        data: {
          bill: {
            title:
              'Authorizes an income tax deduction for the provision of child foster care services',
            openstatesUrl: 'https://openstates.org/mo/bills/2021/HB429',
            updatedAt: '2021-04-17 07:56:42.948232+00:00',
            fromOrganization: {
              name: 'House',
            },
          },
        },
      },
    };
    const component = TestRenderer.create(
      <MockedProvider mocks={[billDetailsMockSuccess]} addTypename={false}>
        <BillDetailsScreen {...billProps} />
      </MockedProvider>
    );
    const loadingScreen = component.toJSON();
    expect(loadingScreen).toMatchSnapshot();

    await wait();

    const successScreen = component.toJSON();
    expect(successScreen).toMatchSnapshot();
  });
  test('Bill details network error screen', async () => {
    const billDetailsMockNetworkError = {
      request: {
        query: gql`
          ${BILL_DETAILS_QUERY}
        `,
        variables: {
          id: 'ocd-bill/2bd26615-0165-4f07-a53e-b9739f53dae6',
        },
      },
      error: new Error('Network error!'),
    };
    const component = TestRenderer.create(
      <MockedProvider mocks={[billDetailsMockNetworkError]} addTypename={false}>
        <BillDetailsScreen {...billProps} />
      </MockedProvider>
    );
    await wait();

    const networkErrorScreen = component.toJSON();
    expect(networkErrorScreen).toMatchSnapshot();
  });
  test('Legislature details graphql error screen', async () => {
    const billDetailsMockGraphQLError = {
      request: {
        query: gql`
          ${BILL_DETAILS_QUERY}
        `,
        variables: {
          id: 'ocd-bill/2bd26615-0165-4f07-a53e-b9739f53dae6',
        },
      },
      result: {
        errors: [new GraphQLError('GraphQL Error!')],
      },
    };
    const component = TestRenderer.create(
      <MockedProvider mocks={[billDetailsMockGraphQLError]} addTypename={false}>
        <BillDetailsScreen {...billProps} />
      </MockedProvider>
    );
    await wait();

    const graphqlErrorScreen = component.toJSON();
    expect(graphqlErrorScreen).toMatchSnapshot();
  });
});
