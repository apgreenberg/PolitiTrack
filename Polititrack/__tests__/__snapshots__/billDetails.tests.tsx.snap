// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bill Detail tests Bill details network error screen 1`] = `
<View>
  <Text
    style={
      Object {
        "fontSize": 12,
        "textAlign": "center",
      }
    }
  >
    Error: 
    Network error!
  </Text>
</View>
`;

exports[`Bill Detail tests Legislature details graphql error screen 1`] = `
<View>
  <Text
    style={
      Object {
        "fontSize": 12,
        "textAlign": "center",
      }
    }
  >
    Error: 
    GraphQL Error!
  </Text>
</View>
`;

exports[`Bill Detail tests bill loading and success screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    Loading
  </Text>
</View>
`;

exports[`Bill Detail tests bill loading and success screen 2`] = `null`;
